### ISSUE TITLE
appimage, flatpak, rpm or archlinux aur

### ISSUE THREAD

**Issue by rezad1393** [openscopeproject/TrguiNG#91]
I wanted to try this app but it doesn't have a x64 version of bianry for linux (not a deb one).

maybe a flatpak or an appimage would help users like me.

**Comment by qu1ck** [comment]
Flatpak is not viable for this app because of it's sandboxing behavior.

AUR package, rpm package, any other package is up to community to contribute because I don't use these systems regularly and can't invest the time to figure out all the packaging specifics of each one.

If someone wants to contribute, make sure you modify the build action to include the new artifact in the automatically generated releases.

But I will look into appimage as it seems that it should work and is at least useful for wide variety of distros.

**Comment by rezad1393** [comment]
I think flatpak is more work than appimage so.

it is just that for flatpak I just use a command to update all of them. but appimage is not easy to update.

btw I use that tremesof app and then added a folder with flatseal to access. so that sandbox is not that much of an issue.

**Comment by qu1ck** [comment]
Ok, so I tried appimage and if I build it on older ubuntu then it packages in buggy webkitgtk which does not render properly. If I build it on newer ubuntu then it is less buggy but will not work on anything with older libs which defeats the portability of the appimage.
I suspect flatpak will have the same fate.

So, for now, self compile it is for linux users.

If anyone wants to experiment with appimage further, literally only change to enable it is to add "appimage" [here](https://github.com/openscopeproject/TrguiNG/blob/5f9f518357c1b001684121996cd1b1fb390b34a1/src-tauri/tauri.conf.json#L16)

**Comment by rezad1393** [comment]
I am confused.

I thought appimage was almost self-contained.
flatpak too.

so if I use new flatpak package on older os (like ubuntu 18) it would still work because it would take most of it libs and files from its own flatpak repo and not host.

maybe I am missing something.

**Comment by qu1ck** [comment]
No, it works with forward compatibility, even then, sometimes not. It almost never works in backwards direction. Appimage will not bundle your whole /usr/lib, it still relies on the fact that your client system will have compatible (which means same or newer) glibc, gtk and other core libs that make up the majority of GNU/Linux. Otherwise the image would be not 80mb but the size of half a linux distro.

Appimage instructions directly advise to build on the oldest possible linux you can support so that everything newer has a good chance to work because for the most part linux tries to maintain strict backward compatibility in core libs.

Unfortunately for TrguiNG even Ubuntu 20.02 which is not that old has buggy webkit so only viable appimage is from 22.04. Chances of it working on anything that is not quite recent are slim.

I'll still include appimage with next release, maybe it will be useful for some.

**Comment by rezad1393** [comment]
I think flatpak has more flexibility with respect to forward/backward versioning.

**Comment by qu1ck** [comment]
I don't see how that is technically possible, you either pack literally every lib and get a 1gb bundle or you don't have backward compatibility, those are the choices.

Anyway, I added an appimage to the last release, it's built on ubuntu 22.

**Comment by rezad1393** [comment]
> Anyway, I added an appimage to the last release, it's built on ubuntu 22.

thank you for the appimage.

but about that

> get a 1gb bundle

you can search the internet and see you answer about flatpak size.
like this:
https://forums.linuxmint.com/viewtopic.php?t=275123

or this quote :
"So how big are these runtimes? On a fresh machine, install KCalc from Flathub. You’re looking at a nearly 900 MB download to get your first runtime. For a calculator."

**Comment by qu1ck** [comment]
Great, so flatpak does indeed bring a whole new OS (sans a kernel) with an app. At least it's smart enough to reuse the runtime between apps when it can.

I looked more into building/publishing a flatpak and it's requirement for fully offline builds and lack of gnome 42 support which has libs needed for tauri introduces a whole host of issues. You can read more about it in https://github.com/tauri-apps/tauri/issues/3619 and linked issues/discussions.
Some folks apparently found workarounds but I couldn't find any clear instructions. 

Looks like this will have to wait until flatpak support is added to tauri (probably in tauri v2 which is currently in alpha).

**Comment by rezad1393** [comment]
I explain to myself flatpak as kinda like docker.

also having read this:
https://theevilskeleton.gitlab.io/2022/05/16/response-to-flatpak-is-not-the-future.html#services
I see this:
"Most AppImage bundles need a big minimum requirement. They don’t bundle glibc or other core dependencies. Additionally, they might not even come with higher level dependencies bundled that would make an app actually functional without relying a lot on the host. Many of them will assume that some dependencies are already installed and have the right version on the host"

as that piece is written by a person defending flatpak model I would assume flatpak has more deps packages in its bundle (or its equivalent).

btw I use flatpak for some apps and it is useful.

for example steam game store. it has a lot of lib32 deps and I didnt want them installed on my PC so I used flatpak.
steam itself is going toward containerization so it seems to work better and better with flatpak (and snap too maybe).

for a single app that usually doesnt need integration, flatpak may be too much and using an appimage for user  is better.

 a good thing about flatpak (that I like) is that app setting and data are in just one folder and unlike other linux apps-systems , they are not scattered everywhere on my home folder. kinda like android but even better.

I would recommend a appimage for users to download and test even if linux distributions later add this program, just so user can test this app before committing to installing it.

**Comment by qu1ck** [comment]
I don't need more convincing that flatpak is useful and more universal. After doing the research about it I was already on the same page.

Currently it's just too hard to package a tauri app into flatpak bundle and write an offline build recipe that one can upload to flathub. Unless I find a (relatively) easy to follow recipe for tauri v1 we will have to wait until tauri v2 and native flatpak support.

**Comment by rezad1393** [comment]
oh.
sorry about the rant.

hope you are successful with build later.

**Comment by qu1ck** [comment]
Note to self: this is a tauri project on flathub that I should be able to model flatpak build after.
https://github.com/flathub/in.cinny.Cinny

**Comment by username227** [comment]
I have never done it, and I don't know if I can. But if nobody else is interested, I might be able to attempt to assist with the AUR. It would be interesting to learn the pkgbuild system. I already have an arch package that was created from your .deb which lists dependencies, etc. Of course, if anybody who has experience is interested in doing this, it would be better. but if not, I can try.

**Comment by username227** [comment]
[PKGBUILD.txt](https://github.com/openscopeproject/TrguiNG/files/13501812/PKGBUILD.txt)

This is my first attempt. This file successfully built and created the package for Arch on my system. rezad, can you test this pkgbuild on your system? download it and remove the .txt extension, and then use the command makepkg. let me know what happens.

**Comment by qu1ck** [comment]
The app is now on flathub https://flathub.org/apps/org.openscopeproject.TrguiNG

**Comment by username227** [comment]
qu1ck - the one thing that won't work well is the desktop file. the app.desktop file in the src-tauri folder is pretty much blank. Is there any way you can incorporate the desktop file that we discussed [here](https://github.com/openscopeproject/TrguiNG/issues/113) in place of the blank one to facilitate a more seamless desktop experience for the end user?

**Comment by qu1ck** [comment]
You can get full desktop file from here
https://github.com/flathub/org.openscopeproject.TrguiNG/blob/master/org.openscopeproject.TrguiNG.desktop

Few notes on your pkgbuild:

* Get db-ip from releases page of this repo. Links from db-ip.com tend to die.
* binary should be executable `install -Dm755`
* License is AGPL-3.0

**Comment by username227** [comment]
OK, definitely do NOT use my PKGBUILD. it crashed my system. i'll try to figure out what's wrong.

**Comment by rezad1393** [comment]
> OK, definitely do NOT use my PKGBUILD. it crashed my system. i'll try to figure out what's wrong.

in span of five hours.
that escalated quickly.

**Comment by rezad1393** [comment]
> Userspace program should not be able to crash your system. I doubt whatever caused your crash is directly caused by you misconfiguring the package, at best it just exposed a hidden problem that was already there.

yeah.
like a graphic driver with amd

**Comment by username227** [comment]
[PKGBUILD.txt](https://github.com/openscopeproject/TrguiNG/files/13514329/PKGBUILD.txt)

As a non-developer, this is a huge learning experience for me. It didn't mess the entire system up per se, but it did really mess up Cinnamon. I believe the problem was related to some missing "" around the install command parameters.

Anyway, this is my second version. I have changed the license, changed the source for the db as you suggested, and changed the permissions (which definitely were an issue). I have not yet worked on the .desktop file issue. That will be next. But this should be good for anybody who wants to test it on an arch-based system.

**Comment by username227** [comment]
OK, so i'm trying to obtain the desktop file directly from the site you mentioned. the command:

`wget https://raw.githubusercontent.com/flathub/org.openscopeproject.TrguiNG/master/org.openscopeproject.TrguiNG.desktop > TrguiNG.desktop`
works just fine, but it saves it with the original filename, not the one I specified.

However, when I put the command into the pkgbuild:

`wget "https://raw.githubusercontent.com/flathub/org.openscopeproject.TrguiNG/master/org.openscopeproject.TrguiNG.desktop" > "$srcdir/TrguiNG.desktop"`

It pulls a file and does not give any errors, but the file is completely empty.

Do you know what's wrong with my commands? Thanks.

**Comment by username227** [comment]
The AUR is now operational. The package is here.

https://aur.archlinux.org/packages/trgui-ng.

**Comment by username227** [comment]
I think I did it. My pleasure. As I said, it was a great learning experience. i'll look into this automation next week unless you get to it first.

**Comment by qu1ck** [comment]
>  i'll look into this automation next week unless you get to it first.

No need, the action requires ssh key saved as secret in github, only I will be able to do that. That's also why I asked to be added as maintainer, so I could use my ssh key.

**Comment by username227** [comment]
I listed you as a maintainer in a commented line on the top of the pkgbuild. let me know if you want any changes to the format of your info.
Also, I successfully uninstalled the package i built manually and installed it directly from the aur using the yay command:
yay -S trgui-ng

On a completely unrelated note - i didn't want to add a whole separate issue for this unless you want me to, but one feature request i'm wondering if it's feasible would be to add a torrent creator engine to the program. Possibly using dottorrent as a base, or something like that.

**Comment by qu1ck** [comment]
> I listed you as a maintainer in a commented line on the top of the pkgbuild. let me know if you want any changes to the format of your info (or you can also).

That file will be autogenerated for each new version and it's not really intended to be viewed by humans anyway so it doesn't matter.

> On a completely unrelated note - i didn't want to add a whole separate issue for this unless you want me to, but one feature request i'm wondering if it's feasible would be to add a torrent creator engine to the program. Possibly using dottorrent as a base, or something like that.

Have you clicked the ![image](https://github.com/openscopeproject/TrguiNG/assets/3622532/be9f18bc-a008-4925-8eb3-d803bea00249) button next to settings?

**Comment by username227** [comment]
> > I listed you as a maintainer in a commented line on the top of the pkgbuild. let me know if you want any changes to the format of your info (or you can also).
> 
> That file will be autogenerated for each new version and it's not really intended to be viewed by humans anyway so it doesn't matter.
> 
> > On a completely unrelated note - i didn't want to add a whole separate issue for this unless you want me to, but one feature request i'm wondering if it's feasible would be to add a torrent creator engine to the program. Possibly using dottorrent as a base, or something like that.
> 
> Have you clicked the ![image](https://private-user-images.githubusercontent.com/3622532/287263092-be9f18bc-a008-4925-8eb3-d803bea00249.png?jwt=eyJhbGciOiJIUzI1NiIsInR5cCI6IkpXVCJ9.eyJpc3MiOiJnaXRodWIuY29tIiwiYXVkIjoicmF3LmdpdGh1YnVzZXJjb250ZW50LmNvbSIsImtleSI6ImtleTEiLCJleHAiOjE3MDE0MzQyMDUsIm5iZiI6MTcwMTQzMzkwNSwicGF0aCI6Ii8zNjIyNTMyLzI4NzI2MzA5Mi1iZTlmMThiYy1hMDA4LTQ5MjUtOGViMy1kODAzYmVhMDAyNDkucG5nP1gtQW16LUFsZ29yaXRobT1BV1M0LUhNQUMtU0hBMjU2JlgtQW16LUNyZWRlbnRpYWw9QUtJQUlXTkpZQVg0Q1NWRUg1M0ElMkYyMDIzMTIwMSUyRnVzLWVhc3QtMSUyRnMzJTJGYXdzNF9yZXF1ZXN0JlgtQW16LURhdGU9MjAyMzEyMDFUMTIzMTQ1WiZYLUFtei1FeHBpcmVzPTMwMCZYLUFtei1TaWduYXR1cmU9ODA0ODBkMTRiNGQ5YzlhNzg5NWIwNmJmNDlkZDMzOGUxNGExZWQzMzk2ZjlkMTMwZDcxZWUxNmE0ZmJlMTQxZCZYLUFtei1TaWduZWRIZWFkZXJzPWhvc3QmYWN0b3JfaWQ9MCZrZXlfaWQ9MCZyZXBvX2lkPTAifQ.UWhZhX_BBqmjOMedjuvhVSSjyCU3Vg9kCeLmU1P-cWI) button next to settings?

LOL. Missed that.... :-)

**Comment by username227** [comment]
> > > I listed you as a maintainer in a commented line on the top of the pkgbuild. let me know if you want any changes to the format of your info (or you can also).
> > 
> > 
> > That file will be autogenerated for each new version and it's not really intended to be viewed by humans anyway so it doesn't matter.
> > > On a completely unrelated note - i didn't want to add a whole separate issue for this unless you want me to, but one feature request i'm wondering if it's feasible would be to add a torrent creator engine to the program. Possibly using dottorrent as a base, or something like that.

OK, makes sense. I was just following the archwiki for headers on the top.

**Comment by username227** [comment]
Perhaps i should figure out how to get the pkgbuild to clean up after itself - delete the git folder that's created to save over 2GB. No reason for a 17MB program to take up 2.2GB.

**Comment by username227** [comment]
FYI, I updated the pkgbuild in the aur with some technical fixes. it will now provide the latest release version, as opposed to the latest commit. If you'd like, I could possibly make a second one that provides the latest build as a -git package, development-type package.

**Comment by qu1ck** [comment]
I just added automation to update AUR repo when new version is released. We'll see how well it works when next release happens, I tested it as far as I could without pushing bogus updates to the actual repo.

> If you'd like, I could possibly make a second one that provides the latest build as a -git package, development-type package

I don't know if there is user demand for this, if you want to add it - go for it. But it should not require automation, probably just remove sha sums checking (if it is allowed) and use "latest" instead of tag in the dbip url. It should just work.

**Comment by qu1ck** [comment]
No, pkgrel is for packaging version only, when software itself changes the version should be updated. That's how version in dev builds works too, it looks like x.x.x-y-g<githash> where y is number of commits since x.x.x tag. That whole string should go into pkgver, but with hyphens replaced with underscores.

**Comment by qu1ck** [comment]
Just published v1.1.0, the AUR update job seems to work correctly. @username227 can you verify that build still works on arch?

**Comment by username227** [comment]
Just tested. Worked like a charm. Great job!

**Comment by qu1ck** [comment]
Great, so now this issue is just about rpm. Any fedora aficionados who would like to pitch in?

**Comment by username227** [comment]
I just added npm as a depencency and changed the pkgrel to 2. Please update on your end if necessary.

**Comment by qu1ck** [comment]
It is not needed at runtime and you already have node which ships with npm in makedepends so the change is redundant.

**Comment by username227** [comment]
nodejs has npm as an optdepends, not makedepends

**Comment by qu1ck** [comment]
Ok, then add it to makedepends, not depends.

**Comment by username227** [comment]
yes, that's what I did since you confirmed it's not needed at runtime, which I suspected. Thanks!

**Comment by username227** [comment]
[error log](https://github.com/openscopeproject/TrguiNG/files/14359387/error.log)
The latest update did not update through the AUR. Attaching copy of error in the build.

**Comment by qu1ck** [comment]
Honestly I can't help much with debugging that build failure because I don't have arch. It complains about being unable to link something from libdbus even though the lib seems to be there.

Possible reasons I can think of:
* dbus is not listed as dependency in the build script. Not sure why it worked before but it should be listed.
* Updated gcc on your system changed behavior of symbol resolution, they planned to make explicit dependency listing mandatory so probably it needs `-ldbus-1` flag or something like that but rust build system or `dbus-rs` build script doesn't provide it for some reason

What does `pkg-config --libs dbus-1` print on your system?

**Comment by username227** [comment]
> What does `pkg-config --libs dbus-1` print on your system?

it outputs: 
-ldbus-1

**Comment by qu1ck** [comment]
As it should, so the lib is there and build system knows how to find it, I'm not sure why it is not on linker the command line. But absolutely nothing in the app code changed between v1.1 and 1.2, not even lib versions. So it must be rust version, gcc version or some other change.
It compiles fine for me on debian 12 with latest rustc.

**Comment by username227** [comment]
The program compiles just fine when I do it manually. Only the script doesn't work. The only differences I can tell in the script are that it has "npm ci" instead of "npm install", but changing that didn't make a difference.
The other difference is the place where we are downloading the mmdb file from. When I do it manually, i am using your wget command on github, but in the script we downloaded it from somewhere else.

**Comment by qu1ck** [comment]
Neither of those are what caused the failure. Probably just dbus is missing as dependency in the script.

> The other difference is the place where we are downloading the mmdb file from. When I do it manually, i am using your wget command on github, but in the script we downloaded it from somewhere else.

I simply mirrored the mmdb file from db-ip.com to release page in this repo because old db files tend to disappear from the original website making old links invalid.

**Comment by username227** [comment]
> so probably it needs -ldbus-1 flag or something like that

How do I add this flag? does it go on the "npm run build -- -b" line like this:??
npm run build -- -b -ldbus-1

**Comment by qu1ck** [comment]
No, that won't work. You probably can add it through defining RUSTFLAGS env variable but you shouldn't have to manually add it, as you say yourself, the manual build works, that means there is some difference in how script environment is setup. Did you try to add dbus to dependencies in the PKGBUILD?

**Comment by username227** [comment]
yes, adding dbus as a dependency didn't make a difference. I checked the diff's between the old pkgbuild and the new one...there's absolutely no reason I can think of why it should work when I do it manually but not when I do it through the pkgbuild. I am stumped. Have to get to work though. Will try to look again later. Hopefully in the interim someone on the aur will yell at us that the pkg is broken and tell us the reason...lol.

**Comment by username227** [comment]
OK, reddit to the rescue. It appears that it won't link properly with LTO enabled. Using options=('!lto') in the package build seems to solve the problem. Not sure why it worked until now, but whatever.

Before I push out a correction on the AUR, someone claimed that git is missing from makedepends. Should I add this? I wanted your input to see if it was necessary or if it's redundant.

**Comment by qu1ck** [comment]
Yes, rust uses git to checkout some libs.

Can you open a PR with the changes for the template here? https://github.com/openscopeproject/TrguiNG/blob/master/ci/generate_pkgbuild.py

**Comment by username227** [comment]
> Yes, rust uses git to checkout some libs.
I don't see git on the list of dependencies for rust in arch, so i'll add it just in case.

> Can you open a PR with the changes for the template here? https://github.com/openscopeproject/TrguiNG/blob/master/ci/generate_pkgbuild.py

Sure, i'll work on that.

**Comment by username227** [comment]
just sharing that everything seems to have worked absolutely flawlessly for the arch package in the latest update!

**Comment by qu1ck** [comment]
1.5.0 now has an rpm package so I think this can be considered complete.
